src/main/scala/edu/berkeley/cs/rise/opaque/sources.scala

```diff
@@ -17,25 +17,46 @@
 
 package edu.berkeley.cs.rise.opaque
 
+import java.io.File
+import java.io.FileInputStream
+import java.io.FileOutputStream
+import java.io.ObjectInputStream
+import java.io.ObjectOutputStream
+
 import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.DataFrame
 import org.apache.spark.sql.SQLContext
 import org.apache.spark.sql.SaveMode
 import org.apache.spark.sql.SparkSession
-import org.apache.spark.sql.DataFrame
 import org.apache.spark.sql.sources.BaseRelation
 import org.apache.spark.sql.sources.CreatableRelationProvider
+import org.apache.spark.sql.sources.RelationProvider
 import org.apache.spark.sql.sources.SchemaRelationProvider
 import org.apache.spark.sql.types.StructType
 
 import edu.berkeley.cs.rise.opaque.execution.Block
 import edu.berkeley.cs.rise.opaque.execution.OpaqueOperatorExec
 
-class EncryptedSource extends SchemaRelationProvider with CreatableRelationProvider {
+class EncryptedSource
+    extends RelationProvider
+    with SchemaRelationProvider
+    with CreatableRelationProvider {
+
+  override def createRelation(
+    sqlContext: SQLContext,
+    parameters: Map[String, String]): BaseRelation = {
+    val schemaFile = new File(parameters("path"), "schema")
+    val schema = new ObjectInputStream(new FileInputStream(schemaFile))
```

Comment:
We should instead use the Hadoop APIs for writing the schema, since `parameters("path")` may be a Hadoop or S3 path.